### PR TITLE
Publish applications

### PR DESCRIPTION
**Pull request by dynamiccast**
Applications can be published by administrators
Administrators can save an image to be the next default image used to boot execution servers.

### CODE REVIEW

File: assets/app/components/single-tab/template.hbs [hunk 1]
```diff
@@ -209,5 +244,63 @@
...
+              <div class="modal-body">
+                <div>
+                  <div>
+                    <div class='centered'>
+                      {{icon-component class='color-success' size=50 materialIcon="done"}}
+                    </div>
+                    <p class='centered color-primary'>Your image has been saved.</p>
+                  </div>
+                  <div class='centered'>
+                    <button type="button" class="btn btn-primary" {{action "saveImagePromptAnswerNo"}}>Aknowledge</button>
```
Comment (dynamiccast) [inline assets/app/components/single-tab/template.hbs]: Aknowledge -> Got it !

**Comment by dynamiccast** [comment]
When file explorer is open, keyboard input should be grabbed out of the VDI.

File: api/hooks/driver.js [hunk 2]
```diff
@@ -39,8 +39,8 @@ function Driver(sails) {
      * @param {Function} callback The initialize callback
      */
     initialize(callback) {
-      sails.after('hook:config:loaded', () => {
-        MachineService.initialize()
+      sails.after('hook:orm:loaded', () => {
+        return MachineService.initialize()
```
Comment (willrnch) [inline api/hooks/driver.js:46]: We need to wait for the `hook:config:loaded` event. `MachineService` uses the `ConfigService`.

File: assets/app/components/file-explorer/component.js [hunk 3]
```diff
@@ -59,21 +67,10 @@ export default Ember.Component.extend({
...
-
   goToDirectory(folder) {
-
-    // removing from current
-    var offset = this.get('history_offset');
-    this.get('history').splice(offset, this.get('history').length - offset);
-
     this.get('history').pushObject(folder);
+    var len = this.get('history').get('length');
+    this.get('history').splice(this.get('history_offset'), (len-1) - this.get('history_offset'));
```
Comment (willrnch) [inline assets/app/components/file-explorer/component.js:73]: What if `len === 0`?

File: assets/app/components/file-explorer/component.js [hunk 4]
```diff
@@ -103,26 +100,25 @@ export default Ember.Component.extend({
...
 
-  pathToString() {
-
-    var data = this.get('history');
-    var offset = this.get('history_offset');
-    var path = '';
+  pathToString: Ember.computed('history', 'history_offset', function() {
+    let data = this.get('history');
+    let offset = this.get('history_offset');
+    let path = '';
```
Comment (willrnch) [inline assets/app/components/file-explorer/component.js]: `path = data.join('\\') + '\\';`

File: assets/app/components/file-explorer/component.js [hunk 5]
```diff
@@ -103,26 +100,25 @@ export default Ember.Component.extend({
...
 
     let name = this.get('selectedFile').get('name').replace(/\.[^/.]+$/, '');
 
-    let m = this.get('store').createRecord('application', {
+    let m = this.get('store').createRecord('app', {
       alias: name,
       displayName: name,
       collectionName: 'collection',
-      path: this.fullPath(),
+      filePath: this.get('pathToString') + this.get('selectedFile').get('name')
```
Comment (willrnch) [inline assets/app/components/file-explorer/component.js]: FYI: You can do `filePath: this.get('pathToString') + this.get('selectedFile.name')`

**Comment by dynamiccast** [comment]
We should be able to cancel image creation.

File: assets/app/components/file-explorer/component.js (hunk 3 above)
Comment (dynamiccast) [inline assets/app/components/file-explorer/component.js:73]: `len` is guaranteed to equal 1 or above. It represents the number of subdirectories you are currently in. Shortest path in the directory tree is `C:\` where `len == 1`.

File: api/controllers/StorageController.js [hunk 6]
```diff
@@ -84,16 +84,27 @@ module.exports = {
...
-      })
-      .catch(res.negotiate);
+    let getFiles;
+
+    if (req.allParams().machines === 'true') {
+      getFiles = MachineService.getMachineForUser(req.user)
+        .then((machine) => {
+          return PlazaService.files({
+            hostname: machine.ip,
+            port: 9090
```
Comment (dynamiccast) [inline api/controllers/StorageController.js]: The port should not be hard coded

**Pull request merged**